packages/gguf/src/gguf.ts

```diff
@@ -0,0 +1,275 @@
+		}
+	}
+}
+
+function readString(view: DataView, offset: number): { value: string; newOffset: number } {
+	const length = view.getBigUint64(offset, true);
+	const value = new TextDecoder().decode(view.buffer.slice(offset + 8, offset + 8 + Number(length)));
+	return { value, newOffset: offset + 8 + Number(length) };
+}
+
+function readMetadataValue(
+	view: DataView,
+	type: GGUFValueType,
+	offset: number
+): { value: MetadataValue; newOffset: number } {
+	switch (type) {
+		case GGUFValueType.UINT8:
+			return { value: view.getUint8(offset), newOffset: offset + 1 };
+		case GGUFValueType.INT8:
+			return { value: view.getInt8(offset), newOffset: offset + 1 };
+		case GGUFValueType.UINT16:
+			return { value: view.getUint16(offset, true), newOffset: offset + 2 };
+		case GGUFValueType.INT16:
+			return { value: view.getInt16(offset, true), newOffset: offset + 2 };
+		case GGUFValueType.UINT32:
+			return { value: view.getUint32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.INT32:
+			return { value: view.getInt32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.FLOAT32:
+			return { value: view.getFloat32(offset, true), newOffset: offset + 4 };
+		case GGUFValueType.BOOL:
+			return { value: view.getUint8(offset) !== 0, newOffset: offset + 1 };
+		case GGUFValueType.STRING:
+			return readString(view, offset);
+		case GGUFValueType.ARRAY: {
+			const arrayType = view.getUint32(offset, true);
+			const arrayLength = view.getBigUint64(offset + 4, true);
+			let arrayOffset = offset + 12;
+			const arrayValues: MetadataValue[] = [];
+			for (let i = 0; i < arrayLength; i++) {
+				const { value, newOffset } = readMetadataValue(view, arrayType, arrayOffset);
+				arrayValues.push(value);
+				arrayOffset = newOffset;
+			}
+			return { value: arrayValues, newOffset: arrayOffset };
+		}
+		case GGUFValueType.UINT64:
+			return { value: view.getBigUint64(offset, true), newOffset: offset + 8 };
+		case GGUFValueType.INT64:
+			return { value: view.getBigInt64(offset, true), newOffset: offset + 8 };
+		case GGUFValueType.FLOAT64:
+			return { value: view.getFloat64(offset, true), newOffset: offset + 8 };
+	}
+}
+
+export type GGUFMetadata = {
+	version: Version;
+	tensor_count: bigint;
+	kv_count: bigint;
+} & Record<string, MetadataValue>;
+
+export interface GGUFTensorInfo {
+	name: string;
+	n_dims: number;
+	shape: bigint[];
+	type: GGMLQuantizationType;
+	offset: bigint;
+}
+
+export interface GGUFParseOutput {
+	metadata: GGUFMetadata;
+	tensorInfos: GGUFTensorInfo[];
+}
+
+export async function gguf(url: string): Promise<GGUFParseOutput> {
+	const r = new RangeView(url);
+	await r.fetchChunk();
+
+	if (r.view.getUint32(0, true) !== new DataView(ggufMagicNumber.buffer).getUint32(0, true)) {
+		throw new Error("not a valid gguf file: not starting with GGUF magic number");
+	}
+
+	const version = r.view.getUint32(4, true);
+	if (!isVersion(version)) {
+		throw new Error("not a valid gguf file: unsupported version");
```

Comment:
```suggestion
		throw new Error(`not a valid gguf file: unsupported version "${version}"`);
```
make it little more descriptive